cirq/testing/sample_circuits_test.py

```diff
@@ -19,5 +19,5 @@ def test_nonoptimal_toffoli_circuit():
     q0, q1, q2 = cirq.LineQubit.range(3)
     cirq.testing.assert_allclose_up_to_global_phase(
         cirq.testing.nonoptimal_toffoli_circuit(q0, q1, q2).to_unitary_matrix(),
-        cirq.TOFFOLI(q0, q1, q2).matrix(),
+        cirq.unitary(cirq.TOFFOLI(q0, q1, q2)),
```

Comment:
Currently it seems that `cirq.unitary(cirq.TOFFOLI(q0, q2, q1))` (or any other permutation of the qubits) gives the same result here. This seems like a bad thing to me. It's not actually possible to understand these matrices, or to correctly combine them with matrices from other operations, unless we know the basis. I don't know the best way to deal with this, but I think it's something we need to reckon with because I could imagine this producing subtle bugs that are hard to track down. Could `cirq.unitary` take an extra argument which is the qubit basis in which to express the matrix? Or always return the matrix of an operation in some canonical basis with the involved qubits sorted? Or return an array that has extra information attached which gives the basis?

Reply:
I don't think that's what we should do, because it would require every implementation of _unitary_ to take a basis. Presumably that's a list of qubits, but some of them don't have any qubits associated. We have to settle for the user using the object itself as a reference for what the basis is. In this case, they would have to combine the matrix with the qubits property.